src/gz/write.rs

```diff
@@ -180,3 +182,177 @@ impl<W: Write> Drop for GzEncoder<W> {
+    /// written to the output stream. After this function is called then further
+    /// calls to `write` may result in a panic.
+    ///
+    /// # Panics
+    ///
+    /// Attempts to write data to this stream may result in a panic after this
+    /// function is called.
+    ///
+    /// # Errors
+    ///
+    /// This function will perform I/O to finish the stream, returning any
+    /// errors which happen.
+    pub fn try_finish(&mut self) -> io::Result<()> {
+        try!(self.inner.finish());
+
+        if self.crc_bytes.len() != 8 {
+            return Err(corrupt())
+        }
+
+        let crc = ((self.crc_bytes[0] as u32) << 0)
+            | ((self.crc_bytes[1] as u32) << 8)
+            | ((self.crc_bytes[2] as u32) << 16)
+            | ((self.crc_bytes[3] as u32) << 24);
+        let amt = ((self.crc_bytes[4] as u32) << 0)
+            | ((self.crc_bytes[5] as u32) << 8)
+            | ((self.crc_bytes[6] as u32) << 16)
+            | ((self.crc_bytes[7] as u32) << 24);
+        if crc != self.inner.get_ref().crc().sum() as u32 {
+            return Err(corrupt());
+        }
+        if amt != self.inner.get_ref().crc().amount() {
+            return Err(corrupt());
+        }
+        Ok(())
+    }
+
+    /// Consumes this decoder, flushing the output stream.
+    ///
+    /// This will flush the underlying data stream and then return the contained
+    /// writer if the flush succeeded.
+    ///
+    /// Note that this function may not be suitable to call in a situation where
+    /// the underlying stream is an asynchronous I/O stream. To finish a stream
+    /// the `try_finish` (or `shutdown`) method should be used instead. To
+    /// re-acquire ownership of a stream it is safe to call this method after
+    /// `try_finish` or `shutdown` has returned `Ok`.
+    ///
+    /// # Errors
+    ///
+    /// This function will perform I/O to complete this stream, and any I/O
+    /// errors which occur will be returned from this function.
+    pub fn finish(mut self) -> io::Result<W> {
+        try!(self.inner.finish());
+        Ok(self.inner.take_inner().into_inner())
+    }
+
+    fn write_buf(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let n = if let Some(Status::StreamEnd) = self.inner.op_status() {
+            0
+        } else {
+            try!(self.inner.write(buf))
+        };
+        if let Some(Status::StreamEnd) = self.inner.op_status() {
+            if n < buf.len() && self.crc_bytes.len() < 8 {
+                let d = cmp::min(buf.len(), n + 8 - self.crc_bytes.len());
+                self.crc_bytes.extend(&buf[n..d]);
+                return Ok(d)
+            }
+        }
+        Ok(n)
+    }
+}
+
+impl<W: Write> Write for GzDecoder<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.header.is_none() {
+            let mut cur = io::Cursor::new(buf);
+            match read_gz_header(&mut cur) {
+                Err(err) => Err(err),
```

Comment:
I'm not sure that this is the semantics that we'll want for this stream unfortunately. If you write half of a valid header then I wouldn't expect the call to `write` to return an error, for example.

Reply:
I am still working on this.

Reply:
added buffer for header bytes. 

state machine would be preferable but I am not ready to implement it :)